[4~9단계 - 방탈출 예약 관리] 라젤(문선영) 미션 제출합니다.

RaZel713 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
아직 부족한 부분이 많은 것 같습니다!
## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [x] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

RaZel713 committed:
docs: 4단계 요구 사항 작성

RaZel713 committed:
feat: spring-boot-stater-jdbc, h2 의존성 추가

RaZel713 committed:
feat: resources에 데이터베이스 테이블 생성 쿼리 작성

RaZel713 committed:
feat: application.properties 설정 추가

RaZel713 committed:
test: 데이터 베이스 적용 테스트 작성

RaZel713 committed:
docs: 5단게, 6단계 요구사항 작성

RaZel713 committed:
test: 테스트 환경 설정

RaZel713 committed:
feat: 데이터 베이스를 활용해 예약 조회 API 처리 로직에서 저장된 예약을 조회

RaZel713 committed:
feat: 데이터 베이스를 활용해 예약 추가/취소 API 구현

RaZel713 committed:
test: 데이터 베이스를 활용한 예약 조회/추가/취소 API 테스트 구현

RaZel713 committed:
refactor: 미사용 메서드 제거

RaZel713 committed:
docs: 7단계 요구사항 작성

RaZel713 committed:
docs: 시간 관리 API 작성

RaZel713 committed:
feat: 시간 관리 페이지 응답 구현

RaZel713 committed:
feat: resources에 예약 시간 데이터베이스 테이블 생성 쿼리 작성

RaZel713 committed:
feat: 시간 추가 API 구현

RaZel713 committed:
feat: 시간 조회 API 구현

RaZel713 committed:
feat: 시간 삭제 API 구현

RaZel713 committed:
test: 시간 관리 API 테스트 추가

RaZel713 committed:
refactor: 클래스명 수정

RaZel713 committed:
docs: 8단계 요구사항 작성

RaZel713 committed:
docs: 예약 관리 API 수정

RaZel713 committed:
feat: 예약 페이지 파일 수정

- templates/admin/reservation-legacy.html 대신 templates/admin/reservation.html 파일을 활용한다.

RaZel713 committed:
feat: 테이블 스키마 재정의

- 외래키 지정을 통해 reservation 테이블과 reservation_time 테이블의 관계를 설정한다.

RaZel713 committed:
refactor: 클래스명 수정

RaZel713 committed:
feat: 예약 시, 시간을 시간 테이블에 저장된 값만 선택할 수 있도록 수정

- 예약 클래스의 시간 타입을 ReservationTime 객체로 수정한다.
- 예약 추가 시, 시간을 문자열 형태로 입력하던 부분을 ReservationTime의 식별자로 수정한다.
- 조회 시 ReservationTime 정보도 함께 조회할 수 있게 쿼리를 수정한다.

RaZel713 committed:
test: 예약 가능 시간이 시간 테이블 기준으로 제한되도록 테스트 수정

RaZel713 committed:
refactor: 생성자 파라미터 정렬

RaZel713 committed:
refactor: 변수명 변경

RaZel713 committed:
refactor: 테스트 순서 및 설명 정리

RaZel713 committed:
refactor: 에러 사인을 상수 처리

RaZel713 committed:
refactor: 주석 삭제

RaZel713 committed:
docs: 9단계 요구사항 작성

RaZel713 committed:
feat: 서비스 레이어 추가

RaZel713 committed:
refactor: 도메인 중심으로 패키지 구조 변경

RaZel713 committed:
test: 계층화 리팩터링 테스트 추가

RaZel713 committed:
docs: README 기능 설명 정리

RaZel713 committed:
refactor: 불변성 보장을 위해 final 키워드 적용

Rok93 commented on src/main/java/roomescape/time/ReservationTime.java:
```diff
@@ -0,0 +1,14 @@
+package roomescape.time;
```
time은 일반적으로 많이 쓰일 수 있는 워딩이다보니 
지금같은 경우에는 도메인 모델의 네이밍과 동일한 패키지명인 reservationtime 과 같이 쓰면 조금 더 도메인 모델들을 구분하기 쉽지 않을까요? 😃

RaZel713 replied:
패키지명을 reservationtime으로 변경하였습니다! 😊

Rok93 commented on src/main/java/roomescape/time/ReservationTimeController.java:
```diff
@@ -0,0 +1,49 @@
+
+    public ReservationTimeController(ReservationTimeService reservationTimeService) {
+        this.reservationTimeService = reservationTimeService;
+    }
```
[빈을 주입하는 방식](https://sumni.tistory.com/146)이 여러개라는 사실 아시나요? 😃
이번 기회에 각 방식에 대해서 간략히 학습해보면 어떨까요? 
가능하면 TIL이나 블로그 등에도 정리해보면 좋을 것 같아요. (혹시 쓰시면 공유해주세요 🙏)

RaZel713 replied:
# 📚 Spring 빈을 주입하는 방식

Spring에서 DI(Dependency Injection)을 적용하는 방법
1️⃣ Constructor Injection (생성자 주입) ✅ 권장
2️⃣ Setter Injection (세터 주입)
3️⃣ Field Injection (필드 주입)

---

## 1️⃣ Constructor Injection (생성자 주입)

``` java
    public class MyClass {
        private final MyDependency myDependency;

        public MyClass(MyDependency myDependency) {
            this.myDependency = myDependency;
        }
    }
```

- 장점:
  - 의존성 누락을 컴파일 타임에 확인할 수 있습니다. (`final` 사용 가능)
  - 불변성이 보장됩니다.
  - 의존성을 명시적으로 전달 받아, 단위 테스트가 용이합니다.
  - 순환 참조를 애플리케이션 구동 시점에 감지 가능합니다.

- 단점:
  - 의존성이 많아질 경우 생성자가 길어질 수 있습니다.

---

## 2️⃣ Setter Injection (세터 주입)

``` java
    public class MyClass {
        private MyDependency myDependency;

        public void setMyDependency(MyDependency myDependency) {
            this.myDependency = myDependency;
        }
    }
```

- 장점:
  - 선택적 의존성 주입 가능합니다.
  - 일부 의존성만 교체하여 테스트 가능합니다.

- 단점:
  - 객체 생성 이후에도 변경 가능합니다. → 단, 불변성이 깨질 수 있습니다.
  - 주입 누락 시, NullPointerException이 발생할 위험이 존재합니다.

---

## 3️⃣ Field Injection (필드 주입)

``` java
    public class MyClass {
        @Autowired
        private MyDependency myDependency;
    }
```

- 장점:
  - 코드가 간결하고 작성하기 쉽습니다.

- 단점:
  - 캡슐화를 위반할 가능성이 생깁니다.
  - 의존성 주입 여부가 명확하지 않습니다.
  - 순환 참조 발생 시 런타임 에러로 확인됩니다.
  - Reflection을 이용한 접근이 필요해 테스트가 번거롭고, 목(mock) 객체 주입이 어렵습니다.

Rok93 commented on src/main/java/roomescape/time/ReservationTimeController.java:
```diff
@@ -0,0 +1,49 @@
+            @RequestBody final ReservationTime reservationTime
+    ) {
+        reservationTimeService.createReservationTime(reservationTime);
+        return ResponseEntity.ok().build();
```
이번 미션에서는 각 API의 status code를 정해준 것 같은데, 
한 번쯤 [status code](https://developer.mozilla.org/ko/docs/Web/HTTP/Reference/Status)에 대해서도 고민해보고 지나가도 좋을 것 같아요. 😃

status code 참고문서를 보신 뒤, 생성 및 삭제에 대한 status code는 어떤 것을 사용하는 것이 좋을지도 간략히 코멘트로 남겨주시면 어떨까 싶어요. 😉

RaZel713 replied:
### • POST 요청을 처리할 때 예약(예약 시간)이 생성되면, 
**201 Created 상태 코드**를 사용하는 것이 좋다고 생각합니다. 
> ➡️**201 Created 상태 코드**: 자원이 성공적으로 생성되었음을 나타내고, 새로운 리소스가 생성되었을 때 클라이언트에게 전달되는 상태 코드

### • DELETE 요청 시, 해당 아이디의 예약(예약 시간)이 존재하지 않으면, 
현재와 같이 **404 Not Found 상태 코드**를 사용하는 것이 좋다고 생각합니다.
> ➡️**404 Not Found 상태 코드**: 클라이언트가 요청한 리소스가 존재하지 않음을 나타내는 상태 코드

### • DELETE 요청이 성공적으로 처리되었을 때,
본문(body)을 반환하지 않으므로 **204 No Content 상태 코드**를 사용하는 것이 좋다고 생각합니다.
> ➡️**204 No Content 상태 코드**: 요청이 성공적으로 처리되었음을 나타내지만, 반환할 내용은 없음을 의미하는 상태 코드

### • 예약 시간이 이미 존재하는 경우
현재와 같이 **409 Conflict 상태 코드**를 사용하는 것이 좋다고 생각합니다. 
> ➡️**409 Conflict 상태 코드**: 클라이언트가 요청한 자원이 이미 존재하고, 중복된 요청이므로 충돌이 발생했음을 나타내는 상태 코드

Rok93 commented on src/main/java/roomescape/time/ReservationTimeController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<ReservationTime>> readAllTie() {
```
메서드에 오타가 있는 것 같네요. 😉

![image](https://github.com/user-attachments/assets/3cb0f18d-f6c7-4995-b5ea-38f7cfd8783b)

RaZel713 replied:
제출 전에 확인했었는데🤔, 왜 발견하지 못했을까요..?🥲
주의해서 보도록 하겠습니다! 감사합니다!

Rok93 commented on src/main/java/roomescape/time/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,51 @@
+        final String sql = "INSERT INTO reservation_time (start_at) VALUES (?)";
+        jdbcTemplate.update(
+                sql,
+                reservationTime.startAt());
```
사소하지만 update 메서드는 한줄로 써도 되지 않을까 싶어요. 👀 
```suggestion
        jdbcTemplate.update(sql, reservationTime.startAt());
```

RaZel713 replied:
update 메서드를 한줄로 작성하여 좀 더 간결하고 가독성있게 변경하였습니다!

Rok93 commented on src/main/java/roomescape/time/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,51 @@
+                            LocalTime.parse(resultSet.getString("start_at"))
+                    );
+                    return time;
+                });
```
`rowMapper` 를 생성하는 부분은 메서드로 추출하면 재활용할 수 있지 않을까요? 🤔

또 각각의 repository는 둘 다, rowMapper를 생성하는 로직이 필요한 것 같은데, 
Repository 인터페이스에 rowMapper 생성 메서드가 필요한 것 같은디, 애초에 추상 메서드로  rowMapper 생성 메서드를 정의해서 하위 구현체는 강제로 rowMapper 메서드 구현을 강제해보는 것도 괜찮지 않을까 싶어요! 😃

RaZel713 replied:
말씀해주신 대로 rowMapper를 메서드로 분리하고, 
`BaseRepository` 인터페이스를 만들어서 rowMapper() 메서드 구현을 강제하는 방향으로 리팩터링하였습니다!
 
코드 중복도 줄고, Repository 구조도 훨씬 깔끔해진 것 같습니다! 
좋은 의견 감사합니다!😊

Rok93 commented on src/test/java/roomescape/reservation/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,177 @@
+    void insertReservationTest1() {
+        // given
+        ReservationTime reservationTime = reservationTimeRepository.findById(1L);
+        final Reservation reservation = new Reservation(DUMMY_ID, "검프",
```
(리뷰와 관계없는 질문!) 테스트 코드에 검프 닉네임이 종종 보이는데, 라젤의 페어는 누구셨나요?  👀

RaZel713 replied:
제 페어는 **부기**라는 크루 였는데, 
**부기**와 제가 모두 레벨 2 **검프** 조여서 **검프**의 이름을 사용하게 되었습니다!

아마 **검프**는 모르실 것으로 생각합니다... ㅎㅎ

Rok93 commented on src/test/java/roomescape/reservation/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,177 @@
+    private ReservationRepository reservationRepository;
+    private ReservationTimeRepository reservationTimeRepository;
+
+    @Autowired
+    private JdbcTemplate reservationJdbcTemplate;
+    @Autowired
+    private JdbcTemplate reservationTimeJdbcTemplate;
```
reservation과 reservationTime를 각각 다루는 jdbcTemplate을 만드실 필요 없이 
하나의 jdbcTemplate만 빈을 주입받아 사용해도 됐을 것 같아요. 😃

JDBC는 특정 테이블에 소속된게 아닌 단순히 데이터베이스와 통신을 위한 API라고 생각해주시면 좋을 것 같아요. 😃

[해당 글](https://tecoble.techcourse.co.kr/post/2023-06-28-JDBC-DataSource/)을 가볍게 읽어보시면 도움이되시지 않을까 싶어요.

RaZel713 replied:
Step 4-9 진행 중 `@SpringBootTest`와 `@JdbcTest`를 혼용해서 사용하면서 발생한 충돌 문제를 해결하기 위해 JdbcTemplate을 두 개로 나누어 사용했습니다.

그렇게 하면 충돌을 피할 수 있을 것 같아서 그렇게 했었는데, 
리뷰어님께서 알려주신 것처럼 JdbcTemplate은 특정 테이블에 국한되지 않고 데이터베이스와 통신하는 도구일 뿐이라, 
여러 테이블을 다룰 때 하나로 사용할 수 있다는 점을 확인했습니다. 

이 부분에 대해서 제가 JDBC를 확실히 학습하지 못해서 잘못 사용한 것 같습니다.

정확하게 짚어주셔서 정말 감사합니다! 😊

Rok93 commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public synchronized List<Reservation> findAllReservations() {
+        final String sql = "SELECT "
+                + "r.id as reservation_id, "
+                + "r.name, "
+                + "r.date, "
+                + "t.id as time_id, "
+                + "t.start_at as time_value "
+                + "FROM reservation as r "
+                + "inner join reservation_time as t "
+                + "on r.time_id = t.id";
```
쿼리같은 경우에는 자바의 [Text Blocks](https://lts0606.tistory.com/583)을 활용해보면 더 깔끔하게 작성이 가능할 것 같아요. 😉

RaZel713 replied:
Text Blocks을 활용하여 작성하였습니다! 
알려주셔서 감사합니다! 😊

Rok93 commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,41 @@
+
+    public void createReservation(final ReservationRequest request) {
+        final ReservationTime time = reservationTimeRepository.findById(request.timeId());
+        final Reservation reservation = new Reservation(
+                0L,
+                request.name(),
+                request.date(),
+                time
+        );
```
DB에 새롭게 생성하려는 예약 객체는 id가 없을텐데요. 
굳이 생성자 파라미터로 이 id를 받을 필요가 있을까 싶은데요. 😃
생성자 혹은 정적 팩토리 메서드 등으로 id를 제외한 값들만 파라미터로 받도록 해보면 어떨까요?

RaZel713 replied:
생성자를 추가하여 id를 제외한 값들만 파라미터로 받을 수 있게 수정하였습니다!

Rok93 commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,25 +1,26 @@
 
 public record Reservation(
-        long id, String name, LocalDate date, LocalTime time
+        long id,
```
id의 경우에 아직 DB에 저장되지 않은 객체라면 id가 없을텐데요. 
현재 없는 케이스는 대다수 `0L`으로 id를 가지게 될 것 같아요. 🤔

null 자체는 안좋다고 생각할 수도 있지만 저는 없는 값을 억지로 다른 값으로 채우기보다는 확실하게 null이라고 두는게 더 낫다고 생각되는 것 같아요. 😃

라젤은 어떻게 생각하시나요?

RaZel713 replied:
처음에 저는 id를 명시적으로 지정하기보다는, **데이터 베이스에서 자동으로 관리**해야 한다고 생각하고, 
**데이터 베이스에 저장되지 않는 객체는 0L의 값을 가져야한다**고 생각하였습니다.

<details>
<summary>id의 값을 데이터 베이스에서 자동으로 관리해야 한다고 생각한 이유</summary>

1️⃣데이터 무결성 유지
수동으로 id를 지정하면 중복된 id가 삽입될 위험이 생기고, 이는 데이터 무결성을 해칠 수 있다 생각합니다.
AUTO_INCREMENT는 데이터베이스가 고유한 값을 자동으로 생성하여 충돌을 방지할 수 있어 무결성을 유지할 수 있다 판단하였습니다.

2️⃣일관성 있는 값 관리
여러 사용자가 동시에 데이터를 삽입하는 상황에서도, 데이터베이스는 일관되게 증가하는 id를 보장할 수 있습니다.
개발자가 직접 관리할 경우, 동시성 문제로 인해 예기치 않은 오류가 발생할 수 있는 가능성이 존재합니다.

3️⃣책임 분리
애플리케이션이 id를 관리하는 책임을 지는 대신, 데이터베이스에게 고유성 관리 역할을 맡김으로써, 코드가 더 단순해지고 오류 가능성도 줄어든다고 생각합니다.

</details>

<details>
<summary>null을 사용하지 않고 0L을 사용한 이유</summary>

레벨 1 과정에서 “null 사용을 지양하라”는 조언을 들었습니다.
실제로 null을 적절히 다루지 못하면,
**내가 의도한 null인지, 의도하지 않은 오류로 발생한 null인지** 구분하기 어려워 디버깅이 힘들다는 점을 경험했습니다.

그래서 레벨 2 Spring을 학습하는 과정에서는, 
데이터베이스에 저장되지 않은 객체의 id 값을 0L로 초기화해 사용했습니다.

> 현재 학습 중인 MySQL, H2 모두 AUTO_INCREMENT 설정된 컬럼에 0을 삽입하면, 0은 무시되고 다음 시퀀스 값이 자동으로 채워집니다.
> 🔹 MySQL 공식 문서
> If you insert 0 into an AUTO_INCREMENT column, the value is treated as if it were NULL, meaning that the next number in the sequence is used.
> 🔹 H2 공식 문서
> An IDENTITY column starts at 1 and increments by 1 by default.

특별히 설정하지 않는 한 AUTO_INCREMENT를 사용한다면, 실제로 사용되는 id의 값은 1L 부터 사용됩니다.
따라서, 0L은 의미가 없는 값으로 사용될 수 있습니다.

</details>

-----------

피드백을 받고, 학습한 결과
**명확성과 이해의 용이성 측면에서 null로 표현하는 것이 더 좋은 선택이라고 판단**하여 null로 리팩터링하였습니다.

### 데이터베이스에 저장되지 않은 객체의 id 표현법
1. _**null로 표현하기 (best practice)**_
• null은 모든 개발자에게 “아직 존재하지 않는 값” 또는 “초기화되지 않은 값”이라는 의미로 명확하게 인식됩니다.
• 따라서 의도를 명확히 표현할 수 있고, 이해하기 쉽습니다.
• 또한, 데이터 베이스에 저장되지 않는 상태로 id를 활용하는 비즈니스 로직이 없기 때문에 NPE가 발생할 가능성이 없습니다.

2. 의미 있는 “미사용 값”을 상수로 지정해 사용하기
• 예를 들어, `-1L`을 미리 정의해서 “아직 저장되지 않은 id”라는 의미로 사용할 수 있습니다.
• 이 경우 객체는 항상 명확한 값을 가지므로, null 관련 오류를 줄일 수 있습니다.
• 다만, 이 방식은 “왜 -1인가?“를 이해하려면 사전 지식이나 주석, 문서가 필요하고,
AUTO_INCREMENT 동작을 알고 있어야 자연스럽게 받아들일 수 있습니다.
• 또한, 데이터 베이스 별로 동작이 다를 수 있기 때문에 
명확히 하려면 보통 NOT NULL, PRIMARY KEY, CHECK (id > 0) 이런 식으로 직접 제약 조건을 추가해서 사용해야합니다.

Rok93 replied:
저장되지 않은 객체의 id를 0으로 두는 것도 나쁘지는 않은 방식이라고 생각해요. 😃
말씀하신 것처럼 일부 데이터베이스 혹은 라이브러리는 0을 DB에 저장되지 않은 객체의 의미로 사용하기도 하기 때문에, 라젤의 방식이 나쁘다고 생각하지는 않아요.  id를 non-nullable하게 만들기에도 좋은 방식이기도하구요. 😃

저도 null을 매우 싫어하기 때문에 null 사용을 지양하라는 편에 1표이기도하고, 실제로 저도 0으로 초기화해서 쓰는 경우도 있어요. 😅 
다만 굳이 생성자로 넣어주지 않아도 될 값을 직접 지정하여 넣어주고있는 것이 불필요한 할당이라고 생각했어요. 😃

2번 방식은 좋은 방식은 아니라고 생각해요. 
기재해주신 것과 같은 문제도있을 수 있지만 함께 일하는 다른 개발자들은 이해하기 어려운 코드가 될 수도 있기 때문이라 생각해요.

Rok93 commented on src/main/java/roomescape/time/ReservationTimeService.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public void deleteReservationTime(final Long id) {
+        if (reservationTimeRepository.delete(id) == 0) {
```
`0`은 매직넘버로 보이는데요. 👀 
적절히 상수로 추출하여 의도를 드러내주면 어떨까요? 🤗

RaZel713 replied:
`DELETE_NO_ROWS_AFFECTED`라고 명명하여 삭제된 데이터가 존재하지 않음을 표현하였습니다!

Rok93 commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,60 @@
+
+    public synchronized int delete(final long id) {
+        final String sql = "DELETE FROM reservation where id = ?";
+        return jdbcTemplate.update(sql, Long.valueOf(id));
```
IntelliJ IDE가 long을 Long으로 Boxing 할 필요가 없다고하고있네요. 😃 
IntelliJ는 꽤 똑똑한 IDE이니 지금처럼 노란줄이 그어져있다면 꼭 한번 확인해보면 도움이 많이 되실거에요. 😉

<img width="870" alt="image" src="https://github.com/user-attachments/assets/8b4fc1ac-ddd9-4457-b9e5-bfc55a5f66fe" />

RaZel713 replied:
알려주셔서 감사합니다! ☺️
불필요한 박싱 없이 사용할 수 있었네요. 앞으로 경고 메시지도 주의 깊게 보겠습니다! 😊

Rok93 commented on src/main/java/roomescape/time/ReservationTimeService.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public void createReservationTime(ReservationTime reservationTime) {
+        reservationTimeRepository.insert(reservationTime);
```
현재는 예약 시간을 중복해서 생성할 수 있는 것 같은데요. 👀
미션 요구사항에서는 따로 언급이 없었던 것 같지만, 생각해보면 동일한 예약시간이 있으면 안될 것 같지 않나요? 🤔

만약 중복된 예약 시간을 생성하지 못하도록 막는 처리가 필요하다면, 라젤은 어떻게 처리하실 것 같으신가요? 😃

RaZel713 replied:
예외 처리 관련 사항이 두번째 미션인 `방탈출 사용자 예약`에 있다고 봐서 
필수적인 부분들만 하려고 하였는데, 해당 부분을 놓친 것 같습니다!

예약과 달리 예약 시간은 동일한 예약 시간이 존재해서는 안된다고 생각합니다!
<details>
<summary>⏰동일한 예약 시간이 존재해서는 안되는 이유</summary>

현재의 예약 시간 관리 기능은 예약 가능한 시간를 미리 설정해두고 설정한 시간 내에서만
선택할 수 있게 하기 때문에 동일한 예약 시간이 존재할 이유가 없습니다.

만약 동일한 예약 시간 입력이 가능해진다면, 예약 관리에 있어 의도치 않는 문제가 발생할 수 있습니다.

</details>

따라서 중복된 예약 시간을 생성하지 못하도록 막는 처리를 `~Service`에서 할 수 있도록 기능을 추가하였습니다!

`ReservationTimeRepository` 에 데이터 베이스에서 중복된 시간의 존재 여부를 확인하는 메서드를 작성하고, 
`ReservationTimeService` 에서 해당 메서드를 호출하여 중복 시간 존재 여부에 따른 예외 처리를 하였습니다.
```java
 public boolean existsBy(LocalTime startAt) {
        final String sql = "SELECT COUNT(*) FROM reservation_time WHERE start_at = ?";
        Integer count = jdbcTemplate.queryForObject(sql, Integer.class, startAt);
        return count != null && count > 0;
    }
```

```java
public ReservationTime createReservationTime(ReservationTime reservationTime) {
        if (isReservationTimeExist(reservationTime)) {
            throw new IllegalArgumentException("[ERROR] 이미 존재하는 예약 시간이 있습니다.");
        }

        return reservationTimeRepository.insert(reservationTime);
}

private boolean isReservationTimeExist(ReservationTime reservationTime) {
        return reservationTimeRepository.existsBy(reservationTime.startAt());
}
```

RaZel713 replied:
### 데이터 베이스에서 중복된 시간의 존재 여부를 확인하는 쿼리를 사용한 이유

처음 중복된 예약 시간을 생성하지 못하도록 막는 처리를 하는 데 있어
`findAllReservationTime`을 호출하여 모든 예약 시간을 가져오고, 스트림을 사용해 중복 여부를 확인하는 방식으로 설계하였는데

`findAllReservationTime`을 사용하여 모든 데이터를 가져오는 방식은 
작은 데이터셋에서는 문제가 없을 수 있지만, 데이터가 많아질수록 성능 저하를 초래할 것이라고 생각하였습니다.

반면, 데이터베이스에서 중복된 시간을 직접 확인하는 쿼리는 필요한 정보만을 빠르게 처리할 수 있어 더 효율적이고 확장성이 높다고 생각하여 위와 같이 작성하였습니다.

Rok93 replied:
말씀하신 것처럼 서비스단에서 예약 시간을 생성하는 flow에서 직접 확인하여 처리하는 방식도 있겠지만, 
이 `start_at` 컬럼에 Unique 제약사항을 추가하여 DB 단에서 막는 것도 고민해볼 수 있을 것 같아요. 😃

RaZel713 replied:
###  `Repository`에서 중복되는 시간의 존재 여부를 확인 후, `Service`에서 확인하여 처리하는 방식
**장점 1)** 예외 메시지나 처리 방식을 개발자가 직접 정의 가능합니다.
**장점 2)** 조건이 변경되거나 추가되는 등의 상황이 발생한다면, 조건에 따라 유연하게 처리가 가능합니다.

**단점 1)** 데이터셋이 커질수록 성능에 부정적인 영향을 미칠 수 있습니다. 
**단점 2)** 데이터셋에 데이터를 추가하거나 수정할 때마다 확인해주어야 합니다.

---

### 데이터 베이스에 Unique 제약 사항을 추가하는 방식 
``` sql
CREATE TABLE reservation_time
(
    id       BIGINT       NOT NULL AUTO_INCREMENT,
    start_at VARCHAR(255) NOT NULL UNIQUE,  -- 유니크 제약 추가
    PRIMARY KEY (id)
);
```
**장점 1)** 추가 쿼리 없이 데이터 베이스 레벨에서 자동으로 중복을 처리하므로 실행 시간이 짧고 효율적입니다.
**장점 2)** 데이터셋에 데이터를 추가하거나 수정할 때마다 로직을 작성하지 않아 코드가 간결해집니다.
**장점 3)** 데이터 베이스의 무결성을 보장할 수 있습니다.

**단점 1)** 복잡한 조건을 처리하는데 있어 유연성이 부족합니다.
**단점 2)** “duplicate key error”와 같은 기본적인 오류 메시지가 반환되므로, 사용자 친화적으로 처리하려면, 데이터베이스 오류를 잡아서 처리해야합니다.

---

현재 상황에서 위와 같이 장단점을 고려하면
`start_at` 컬럼에 Unique 제약사항을 추가하여 DB 단에서 막는 것이 더 적합하다는 생각이 듭니다.

그런데 **'예약 시간이 중복되서는 안된다'** 라는 조건은 비즈니스 로직의 조건인데,
데이터 베이스에서 이를 알고 처리하는 게 적합한 책임이 맞는지 의문이 듭니다!

로키의 의견을 여쭈어봐도 될까요?

Rok93 replied:
말씀하신 것처럼 ` '예약 시간이 중복되서는 안된다' 라는 조건은 비즈니스 로직의 조건` 이기 때문에 서비스 단에서의 예외처리도 꼭 해줘야한다고 생각해요.  다만 DB의 제약조건도 동일하게 걸어주는 것이 좋다고 생각합니다. 

서비스 단의 유효성 검증은 논리적인 검증이고 DB 단의 제약조건은 물리적인 검증(?)이라고 생각해주면 될 것 같아요. 
서비스단에 유효성만 잘 처리하고있더라도 어느정도는 방어가 되겠지만, 예약 등록 API요청이 너무 많이 들어오게되면 순간적으로 이 서비스 단의 유효성 검증을 뚫고 DB에 중복 데이터를 저장할 수도 있어요. 
단순히 저장만되면 다행이지만, 애플리케이션에서는 이 데이터가 중복될 일이 없다고 가정하여 코드를 짤 확률이 높다고 생각해요. 😅
의도치 않은 결과로 인해서 아예 데이터 조회가 안되고 그로인해 애플리케이션이 정상적으로 동작하지 않을 수 있어요.

Rok93 commented on src/main/java/roomescape/time/ReservationTimeController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<ReservationTime> createTime(
```
예약 시간 관리화면에서 예약 시간을 생성해봤는데요. 
화면은 변경이 없는 것 같아요. 🤔
어떤 문제인지 스스로 해결해보시고 혹시라도 원인을 잘 모르겠다면 따로 DM 등으로 질문 주셔도 좋아요. 😉

<img width="1920" alt="image" src="https://github.com/user-attachments/assets/bd833bc4-ff37-4523-bed4-4aa3f10e3435" />


가능하다면 어떻게 단계적으로 접근했는지도 공유해주시면 좋을 것 같아요. 😃

RaZel713 replied:
### <예약 시간, 예약 생성 시 화면 미갱신 오류 수정 과정>
1. 예약 시간 생성 시 바로 화면에 반영되지 않고 페이지 새로 고침을 해야 반영되는 현상을 확인
2. 생성에는 문제가 없지만, 생성된 값이 제대로 반환되지 않는다고 판단하여 `Controller`를 확인
3. `Controller` 메서드를 확인해본 결과 `ResponseEntity<ReservationTime>`를 반환해야하는데, `ResponseEntity.ok().build()` 를 반환하는 것을 확인
4. 데이터 베이스에 저장한 후 저장된 값을 받아 반환해야 하므로, `Service`와 `Repositiory`에서 저장한 후 저장한 값을 불러와서 반환할 수 있도록 변경
5. 예약 시간 생성, 예약 생성 시 화면이 갱신됨을 확인

### 저장 후, 저장된 값을 불러오는 방법
데이터베이스에서 id 값을 자동으로 관리하기 때문에, 
저장이 완료된 후에는 해당 id를 이용해 저장된 값을 조회해야 정확한 객체를 반환할 수 있습니다.

이를 구현하는 방법에는 

> 1. 가장 마지막에 저장된 id 값을 조회한 뒤, 해당 id로 데이터를 다시 불러오는 방법
> 2. 저장하는 과정에서 KeyHolder를 사용해, 저장된 객체의 id 값을 직접 받아오고, 이를 이용해 데이터를 조회하는 방법

이 있는데, 이중 2번을 사용하여 구현하였습니다.

``` java
 @Override
    public synchronized ReservationTime insert(final ReservationTime reservationTime) {
        final String sql = "INSERT INTO reservation_time (start_at) VALUES (?)";
        KeyHolder keyHolder = new GeneratedKeyHolder();

        jdbcTemplate.update(connection -> {
            PreparedStatement ps = connection.prepareStatement(sql, new String[]{"id"});
            ps.setString(1, reservationTime.startAt().toString());
            return ps;
        }, keyHolder);

        Long id = Objects.requireNonNull(keyHolder.getKey()).longValue();
        return findById(id);
    }
```

<details>
<summary>2번을 사용한 이유</summary>

가장 마지막에 저장된 id 값으로 불러오는 방법은,
동시에 여러 사용자가 저장 요청을 보낼 경우 다른 사용자의 id를 참조하게 될 위험이 있다는 단점이 있습니다.
따라서 여러 사용자가 동시에 접근할 가능성이 있는 경우, 이 방법은 신뢰할 수 없습니다.

반면, KeyHolder를 사용하여 저장할 때 바로 생성된 id 값을 얻어오는 방법은
동시성 문제 없이 올바른 id 값을 받아올 수 있기 때문에 더 안전하고 정확합니다.

</details>

Rok93 reviewed (changes requested):
라젤 4~9단계도 잘 진행해주셨어요. 👍👍 
몇몇 코멘트 남겨두었는데, 확인 부탁드릴게요. 🙏
현제 예약 시간 관리 화면에서 예약 시간을 추가했음에도 예약 시간 화면이 변경되지 않는 이슈가 있는데, 
어떤 문제가있는지 확인하고 수정해주시면 좋을 것 같아요. 😃

RaZel713 committed:
refactor: 패키지명 변경

- 예약 시간 패키지를 'time'에서 'reservationtime'으로 변경한다.

KoSeonJe commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,25 +1,26 @@
+        ReservationTime time
 ) {
 
+    private static final String ERROR_SIGN = "[ERROR] ";
```
공통된 에러 표시를 Reservation 필드 내 상수로 분리해주셨는데, 해당 에러 표시로 에러 메시지를 시작하는 커스텀 클래스를 만들어 보는 것은 어떤가요!?

반복되는 상수 메시지 호출도 제거할 수 있을 것 같아요!

KoSeonJe replied:
공통된 ERROR_SIGN를 Reservation 필드 내 상수로 분리해주셨는데, 해당 ERROR_SIGN로 에러 메시지를 시작하는 커스텀 예외 클래스를 만들어 보는 것은 어떤가요!?

반복되는 상수 메시지 호출도 제거할 수 있을 것 같아요!

KoSeonJe commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -15,29 +15,34 @@
     @PostMapping
     public ResponseEntity<Reservation> createReservation(
-            @RequestBody final Reservation reservation
+            @RequestBody final ReservationRequest request
```
요청 DTO 명을 좀 더 구체적으로 지어줘도 좋을 것 같아요!
예약을 생성해야하는 요청뿐만 아니라 다양한 요청이 있을 수 있다고 생각합니다.

KoSeonJe replied:
요청 DTO 명을 좀 더 구체적으로 지어줘도 좋을 것 같아요!
예약을 생성해야하는 요청뿐만 아니라 다양한 요청이 있을 수 있다고 생각합니다.

KoSeonJe commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public void deleteReservation(final Long id) {
+        if (reservationRepository.delete(id) == 0) {
+            throw new NoSuchElementException("[ERROR] 해당 ID의 예약을 찾을 수 없습니다. id:" + id);
+        }
```
delete의 내부 구현을 알지 못하면 이해하기 힘들 수 있을 것 같아요!
분리해주는 것은 어떤가요?
```suggestion
        int deletedCount = reservationRepository.delete(id)
        if (deletedCount == 0) {
            throw new NoSuchElementException("[ERROR] 해당 ID의 예약을 찾을 수 없습니다. id:" + id);
        }
```

KoSeonJe commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,60 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public synchronized void insert(final Reservation reservation) {
```
synchronized 키워드를 사용하신 이유가 있나요?! 어떤 문제점이 있을까요?
이보다 더 괜찮은 해결 방법은 없을까요?

KoSeonJe replied:
현재 학습 주제인 Spring보다 Database와 관련된 내용이지만..!
synchronized 키워드를 사용하신 이유가 있나요?! 어떤 문제점이 있을까요?
이보다 더 괜찮은 해결 방법은 없을까요?
한번 고민해보시면 좋을 것 같아요!

KoSeonJe commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,41 @@
+        return reservationRepository.findAllReservations();
+    }
+
+    public void deleteReservation(final Long id) {
+        if (reservationRepository.delete(id) == 0) {
+            throw new NoSuchElementException("[ERROR] 해당 ID의 예약을 찾을 수 없습니다. id:" + id);
+        }
+    }
```
delete의 내부 구현을 알지 못하면 이해하기 힘들 수 있을 것 같아요!
분리해주는 것은 어떤가요?
```suggestion
        int deletedCount = reservationRepository.delete(id)
        if (deletedCount == 0) {
            throw new NoSuchElementException("[ERROR] 해당 ID의 예약을 찾을 수 없습니다. id:" + id);
        }
```

KoSeonJe commented on src/main/java/roomescape/time/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,51 @@
+        return jdbcTemplate.update(sql, Long.valueOf(id));
+    }
+
+    public ReservationTime findById(final long id) {
+        final String sql = "SELECT id, start_at FROM reservation_time WHERE id = ?";
+        return jdbcTemplate.queryForObject(sql, (resultSet, rowNum) ->
+                new ReservationTime(
+                        resultSet.getLong("id"),
+                        LocalTime.parse(resultSet.getString("start_at"))
+                ), id
+        );
```
queryForObject메서드 자체적으로 찾은 값이 없다면 예외를 발생시키는 것으로 알고 있습니다! 

저는 값을 찾지 못해 커스텀한 예외를 발생시켜 클라이언트에게 메시지를 보내줘야 한다고 생각하는데, 이것을 repository내 책임으로 볼지도 고민해보면 좋을 것 같아요!

KoSeonJe reviewed (commented):
안녕하세요 라젤!
크루에게 해보는 첫 리뷰네요! ㅎㅎ
적절하지 않은 피드백도 있을 수 있으니 비판적으로 바라봐주세요!

RaZel713 committed:
refactor: 예약 시간 조회 메서드명 수정

- 'readAllTie'에서 'readAllTime'으로 오타 수정

RaZel713 committed:
refactor: 불필요한 Long Boxing 제거

RaZel713 committed:
refactor: 불필요한 줄바꿈 정리

RaZel713 committed:
refactor: Text Blocks을 사용해 쿼리 작성

RaZel713 committed:
fix: 예약 시간 생성 후 화면 미갱신 오류 수정

RaZel713 committed:
fix: 예약 생성 후 화면 미갱신 오류 수정

RaZel713 committed:
feat: 동일한 예약 시간 생성 방지 기능 구현

RaZel713 committed:
refactor: 매직 넘버 상수 처리

KoSeonJe commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,60 @@
+        return jdbcTemplate.update(sql, Long.valueOf(id));
+    }
+
+    public synchronized List<Reservation> findAllReservations() {
```
현재 학습 주제인 Spring보다 Database와 관련된 내용이지만..!
조회할 때 synchronized 키워드를 사용하신 이유가 있나요?! 어떤 문제점이 있을까요?
한번 고민해보시면 좋을 것 같아요!

RaZel713 committed:
refactor: 예약 생성 요청 DTO 명명 변경

RaZel713 committed:
refactor: rowMapper 메서드를 인터페이스로 추출하여 각 Repository에서 구현

RaZel713 committed:
refactor: 패키지 구조 변경

RaZel713 committed:
refactor: Repository의 공통 메서드를 인터페이스로 추출

RaZel713 committed:
refactor: Reservation 생성자 추가

RaZel713 committed:
refactor: 데이터 베이스에 저장되지 않는 예약의 Id를 null 처리

RaZel713 committed:
refactor: id 타입을 원시타입인 long에서 참조타입인 Long으로 변경

- 초기화 되지 않는 id의 값을 null로 표현한다.
- 따라서, id 타입을 참조 타입인 Long으로 변경한다.

RaZel713 committed:
test: 테스트 패키지 구조 변경

RaZel713 committed:
test: ReservationRepositoryTest의 JdbcTemplate을 하나의 JdbcTemplate로 통합

RaZel713 committed:
feat: 중복된 예약 시간 생성시, 409 Conflict 상태 코드를 반환

RaZel713 reviewed (commented):
안녕하세요~ 로키 리뷰어님! 
리뷰를 너무 잘 남겨주셔서 학습하는 데 정말 큰 도움이 되었습니다! 
꼼꼼하게 리뷰해주셔서 정말 감사해요~ ☺️

빠르게 답을 드리고 싶은데, 조금 더 나은 답을 찾고 싶어서 시간이 많이 걸리더라고요 🥲
더 많이 리뷰어님께 피드백 받고 싶은데 혹시 한번 더 리뷰를 부탁드려도 괜찮을까요..?
조심스럽게 부탁드려봅니다!

saera-yook commented on src/main/java/roomescape/Repository/ReservationRepository.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Override
+    public synchronized List<Reservation> findAll() {
```
저는 `synchronized` 키워드에 대해 잘 몰라서 궁금한 것을 질문해도 될까요~?
해당 키워드를 사용하는 이유와 사용 기준에 대해 궁금합니다ㅎㅎ

Rok93 replied:
`synchronized` 키워드는 이번 단계에서는 제거해보면 어떨까 싶어요. 😃

`synchronized` 키워드는 성능이 느려지고 한계점도있기 때문에 사용하는 것을 지양하는 편이에요. 👀 
조금 더 깊이 다루기에는 이번 미션을 너무 벗어나는 이야기인 것 같아서, 이번엔 단순히 제거만 요청드릴게요. 🙏
미션 이후에 시간 여유가 생기면 `synchronized의 한계` 같은 키워드로 관련 내용들을 찾아보시면 좋을 것 같아요.

RaZel713 replied:
로키 리뷰 답 쓰기

RaZel713 replied:
`synchronized` 키워드를 제거 하였습니다!

saera-yook commented on src/main/java/roomescape/Repository/ReservationRepository.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Override
+    public synchronized int delete(final Long id) {
```
`findById()`처럼 `deleteById()`라고 하면 의도가 더 분명하고 통일성도 있는 이름이 될 것 같아요 :)

RaZel713 replied:
id를 사용하는 메서드의 네이밍을 ~By(Long id) 형태로 변경했습니다!

`deleteBy(Long id)`, `findBy(Long id)` 

좋은 리뷰 감사합니다! 😊

saera-yook commented on src/main/java/roomescape/Repository/ReservationRepository.java:
```diff
@@ -0,0 +1,101 @@
+
+        Long id = Objects.requireNonNull(keyHolder.getKey()).longValue();
+
+        return findById(id);
```
`insert()` 메서드 안에 조회 로직까지 포함되어 있네요. 단일 책임 원칙이나 [CQRS](https://velog.io/@suzhanlee/CQRS-%EC%9D%B4%ED%95%B4%ED%95%98%EA%B8%B0) 관점에서 보면 id만 반환하는 것이 좋을 것 같은데 어떻게 생각하시나요~?

saera-yook commented on src/test/java/roomescape/reservation/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,177 @@
+        assertThat(reservationRepository.delete(1L)).isEqualTo(1);
+    }
+
+    @DisplayName("존재하지 않는 아이디가 들어오면 예외가 발생한다.")
+    @Test
+    void removeReservationTest2() {
+        // when & then
+        assertThat(reservationRepository.delete(1L)).isEqualTo(0);
+    }
```
DisplayName을 보면 예외가 발생하는 상황을 테스트하는 메서드라고 생각되는데 실제로는 0을 반환하는지 검증되고 있네요~ DisplayName을 수정하거나 검증 방식을 수정하면 어떨까요~?

kwonkeonhyeong commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -10,34 +10,42 @@
+import roomescape.domain.reservation.Reservation;
+import roomescape.service.ReservationService;
 
 @Controller
```
현재 '@Controller'와 '@RestController' 중에서 '@Controller'를 사용하고 있는 이유가 있나요?? 

'@Controller'를 선택하게 된 라젤의 생각이 궁금해요🤗

saera-yook commented on src/test/java/roomescape/reservation/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,177 @@
+        assertThat(reservationRepository.findAll()).hasSize(1);
+    }
+
+    @DisplayName("예약이 저장되면 아이디를 붙인 예약을 반환한다.")
+    @Test
+    void insertReservationTest2() {
+        // given
+        ReservationTime reservationTime = reservationTimeRepository.findById(1L);
+        final Reservation reservation = new Reservation(DUMMY_ID, "검프",
+                LocalDate.of(2025, 4, 4), reservationTime);
+
+        // then
+        assertThatCode(() -> reservationRepository.insert(reservation))
+                .doesNotThrowAnyException();
+    }
```
예약 저장 후 반환된 예약이 id를 갖고 있는지 검증되어야 할 것 같은데 저장 시 예외가 발생하지 않는지만 검증되고 있네요~
이 검증 로직은 `insertReservationTest1()`에서 검증되면 더 좋을 것 같아요 :)

saera-yook reviewed (commented):
라젤 안녕하세요~
저도 첫 크루 리뷰를 라젤에게 남기게 되었네요ㅎㅎ 
해당하지 않는 id로 삭제 요청 시 예외 처리 부분과 테스트 클래스를 세분화 한 것이 인상깊었어요👍
워낙 잘 구현하셔서 크게 말씀드릴 부분이 없어 질문과 간단한 코멘트들만 남겼습니다!
혹시 제가 잘못 이해한 부분이 있다면 알려주세요😃 그럼 스터디 때 만나요🥰

kwonkeonhyeong commented on src/main/java/roomescape/domain/reservation/Reservation.java:
```diff
@@ -0,0 +1,30 @@
+        ReservationTime time
+) {
+
+    private static final String ERROR_SIGN = "[ERROR] ";
```
에러 메시지 내에서 '[ERROR]' 가 공통적으로 사용하기 때문에 이를 상수로 처리하였다고 생각하는데, 
Reservation 객체 이외은 다른 객체에서는 `[ERROR]`를 사용하지만 이를 상수 처리하지 않은 객체도 존재하여 일관성이 깨진하고 생각해요!

그리고 만약 예외 메시지의 prefix가 변경된다면 예외 메시지를 사용하는 모든 객체를 돌아다니며 수정해줘야 하는데 이런 부분에 대해서는 어떻게 생각하시는지 라젤의 생각이 궁금합니다 🧐

kwonkeonhyeong commented on src/test/java/roomescape/reservation/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,177 @@
+        assertThat(reservations.size()).isEqualTo(count);
+    }
+
+    @DisplayName("육단계: 데이터 추가 / 삭제하기")
```
오단계, 육단계, 팔단계에 대한 테스트를 RepositoryTest에게 맡기는 것이 어색하다고 생각했습니다!
위 테스트들은 Repository Test 라고 생각되기 보다는 요청 응답에 대한 테스트에 가까워 Controller Test에서 진행하는 것이 적절하다는 생각이 드는데, 라젤의 생각은 어떤가요?

Rok93 commented on src/main/java/roomescape/domain/reservation/CreateReservationRequest.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.time.LocalDate;
+
+public record CreateReservationRequest(
```
DTO는 별도의 패키지로 분리해서 두면 어떨까요?

RaZel713 replied:
DTO를 별도의 패키지로 분리하였습니다! 👍

Rok93 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,42 @@
+    public void deleteReservation(final Long id) {
+        if (reservationRepository.delete(id) == DELETE_NO_ROWS_AFFECTED) {
+            throw new NoSuchElementException("[ERROR] 해당 ID의 예약을 찾을 수 없습니다. id:" + id);
+        }
```
Service 계층에서는 (DB에 대한 세부적인 사항보다는) 단순히 Repository 객체에게 삭제하라는 메시지를 보내서 성공했는지 실패했는지 여부만 궁금하지 않을까요? 😃

지금의 경우에는 DB와 소통하는 것은 Repository 계층인데, Service 계층이 이러한 세부 사항들을 알아야하는 것 같아요.

RaZel713 replied:
예약 삭제 및 예약 시간 삭제 기능에서 
Repository가 삭제된 행 수 대신 삭제 성공 여부를 반환하고, 
Service가 성공 여부에 따라 예외를 던질 수 있게 수정하였습니다.

이처럼 Service 계층이 Repository의 세부 구현(삭제된 행 수를 반환)에 의존하지 않도록 개선하였습니다.

알려주셔서 감사합니다. 😊

Rok93 commented on src/main/java/roomescape/Repository/ReservationRepository.java:
```diff
@@ -0,0 +1,101 @@
+                """;
+
+        return jdbcTemplate.query(
+                sql, (resultSet, rowNum) -> mapRow(resultSet));
```
`(resultSet, rowNum) -> mapRow(resultSet)` 람다식 자체인 rowMapper를 메서드로 추출해보면 어떨까요?  
메서드 시그니처를 잘 모르겠다면... 추출하려는 부분을 드래그한 뒤, 단축키를 활용해봐도 좋을 것 같아요! 😃 


### 메서드 추출 단축키 
* MAC: `⌥(Option) + ⌘(Command) +  M(Method)`
* WINDOW: Ctrl + Alt + M  (정확히는 기억이 안나는데 아마도 이게 맞을꺼에요. 👀)

RaZel713 replied:
가독성 향상과 코드 재사용을 위해 rowMapper를 메서드로 추출하였습니다! 😊

Rok93 commented on src/main/java/roomescape/Repository/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public boolean existsBy(LocalTime startAt) {
+        final String sql = "SELECT COUNT(*) FROM reservation_time WHERE start_at = ?";
```
SQL의 `EXISTS`를 활용해서 true/false 결과를 바로 얻을 수도 있지 않을까요? 😃

Rok93 replied:
만약 COUNT 쿼리를 사용할거라면 `LIMIT 1` 제약을 걸어서 1개의 row만 읽어올 수 있도록 조건을 추가해준다면 조금 더 쿼리 성능이 좋아질 것 같아요. 

다만 성능이 좋아진다고 말씀드린 전제조건은 `조건에 부합하는 row가 많은 경우`일거라 서비스단에서 예약시간 중복 생성을 제한을 하고있기 때문에 문제 상황이 아니라면 보통은 최대 1건만 조회가 될 것 같네요. 🤔

RaZel713 replied:
``` java
 public boolean existsBy(LocalTime startAt) {
        final String sql = "SELECT EXISTS(SELECT 1 FROM reservation_time WHERE start_at = ?)";
        Boolean exists = jdbcTemplate.queryForObject(sql, Boolean.class, startAt);
        return Boolean.TRUE.equals(exists);
}
```
위와 같이 SQL의 `EXISTS`를 활용해서 true/false 결과를 바로 얻을 수 있도록 변경하였습니다.

<details>
<summary> EXISTS를 사용하는 것이 더 나은 이유 </summary>

**COUNT(*)** 는 모든 레코드를 세야하는 반면,
**EXISTS** 는 하나라도 찾으면 바로 멈추기 때문에 더 빠르고 효율적입니다.
즉, 데이터 베이스 입장에서 리소스가 절약됩니다.

</details>

<details>
<summary>SELECT 1를 쓴 이유</summary>

EXISTS는 안에 든 쿼리가 결과가 있으면 true, 없으면 false를 반환합니다.
해당 메서드는 주어진 예약 시간이 데이터 베이스에 존재하는 지 여부를 확인하므로, 
가장 작은 수인 1을 사용하여 존재 여부만 빠르게 확인할 수 있도록 하였습니다.

</details>

<details>
<summary>queryForObject를 쓴 이유</summary>

`query`는 직접 ResultSet을 받고, row마다 수동으로 매핑해야하는 반면, 
`queryForObject`는 한 건 조회 + 타입 변환을 자동으로 해주기 때문에 
'한건 조회' 할 때에는 훨씬 간결하고 실수 없이 사용할 수 있습니다.

</details>

#### 추가학습) COUNT를 사용한다면..?
`final String sql = "SELECT COUNT(1) FROM reservation_time WHERE start_at = ? LIMIT 1";`

- COUNT(1): 실제로 데이터가 존재하면 1을 세고, 없으면 0을 셉니다. (존재 여부만 체크)
- LIMIT 1: WHERE 조건에 맞는 첫 번째 레코드만 조회하고, 더 이상 읽지 않습니다. 
                 만약 조건에 맞는 row가 많은 경우, 처음 하나만 세고 끝낼 수 있습니다.

Rok93 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -10,34 +10,42 @@
+import roomescape.domain.reservation.Reservation;
+import roomescape.service.ReservationService;
 
 @Controller
```
[라젤이 이전에 남겨주신 답변](https://github.com/woowacourse/spring-roomescape-admin/pull/217#discussion_r2051508289)을 되짚어봤을 때, 이 컨틀롤러는 `@RestController`로 변경해볼 수도 있지 않을까요? 😃

RaZel713 replied:
`@Controller`를 `@RestController`로 변경하여 JSON 응답 컨트롤러임을 명확히 하였습니다.

상태 코드에 따라 `ResponseEntity`을 유지하거나 제거하였습니다.

Rok93 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -10,34 +10,42 @@
     }
 
     @GetMapping
     public ResponseEntity<List<Reservation>> readAllReservation() {
```
요청이나 응답은 가능한 DTO를 반환해주면 어떨까요? 😃 
지금은 요구사항이 간단하고 응답하는 Reservation 요소의 응답 포맷이 모두 동일하기 때문에 문제가 없지만, 
추후에는 각 API들이 Reservation 요소를 응답하지만 응답 포맷이 다를 수도 있어요. 
또한 고객 정보 같은 User를 조회하는 경우 `password`와 같이 노출되면 안되는 정보까지 내려주기가 너무 쉬워질 수 있죠. 😱😨

이런 포인트를 고려했을 때, 요청이나 응답은 DTO로 변환해서 내보내는 것이 조금 더 안정적이고 또 다양한 API 요구사항을 만족하기가 수월해지는 것 같아요. 

아래의 글을 참고해보시면 좋을 것 같아요. (지금의 단계에서는 Entity는 도메인 객체 정도로 생각해주셔도 될 것 같아요)
[요청과 응답으로 엔티티(Entity) 대신 DTO를 사용하자](https://tecoble.techcourse.co.kr/post/2020-08-31-dto-vs-entity/)

RaZel713 replied:
요청과 응답에 있어 DTO를 활용하게끔 리팩터링 하였습니다!😊

질문) 아래와 같이 DTO를 사용하였는데, 제대로 활용한 것이 맞는지 여쭈어봐도 될까요?

**`Controller`  ➡️  `Service`**   : Request (DTO)
 
**`Service`**                               :  Request (DTO) ➡️ Reservation 

**`Service` ↔️ `Repository`** : Reservation

**`Service`**                               :  Reservation ➡️ Response (DTO) 
 
**`Service`  ➡️  `Controller`**   : Response (DTO)

Rok93 replied:
DTO 객체를 어디까지 사용할 것인지에대해서 혹은 DTO 객체를 어디서 생성해줘야하는지에 대해서는 명확한 답은 없다고 생각해요. 😃
DTO 객체 생성은 어디에서 해야하는지? 에 대해서 크루들과 가볍게 논의해보면 어떨까 싶어요. 💬

토론한 내용 혹은 토론 이후 라젤의 생각을 먼저 남겨주신다면 제 의견도 남겨두겠습니다. 🙏 
(제가 먼저 남기면 그걸 정답처럼 생각할까봐 먼저 논의해보시면서 고민해보면 좋을 것 같아 토론을 권장드려봅니다. 😉)

RaZel713 replied:
### Controller에서 생성하는 경우
- 장점) 서비스가 도메인 로직 만을 처리하므로, 테스트가 쉬워지고 서비스의 활용도가 높아집니다.
- 단점) 컨트롤러는 입출력 계층이므로, 도메인 객체나 비즈니스 로직과 직접 연관된 DTO 생성 책임을 갖는 것은 역할 분리에 어긋날 수 있습니다.

### Service에서 생성하는 경우
- 장점) 비즈니스 흐름 내에서 필요한 데이터 변환과 로직을 한 곳에서 관리할 수 있어 일관성 있는 처리가 가능합니다.
- 단점) 서비스는 도메인 로직을 조율하는 계층인데, DTO 생성까지 맡게 되면 책임이 많아져 복잡도가 증가할 수 있습니다.

### Assembler 혹은 Mapper를 활용하는 경우
- 장점) DTO 생성을 별도 클래스로 분리하면 관심사를 명확하게 나눌 수 있고, 테스트나 유지보수에도 유리합니다.
- 단점) 작은 규모의 프로젝트에서는 과한 설계가 될 수도 있습니다. 또한 클래스 수가 늘어나고 설정할 게 많아져서 초기 학습 비용과 구현 비용이 커질 수 있습니다.

로키가 말씀해주신 것처럼 DTO 사용에 대한 정답은 없다고 생각합니다!

각 방식에는 장단점이 존재하기 때문에, 상황에 따라 적절히 선택하는 것이 중요하다고 느꼈습니다.
현재 미션처럼 작은 단위의 프로젝트에서는 프로그램의 흐름에 맞게 작성해도 좋을 것 같다는 생각이 들었습니다!

Rok93 reviewed (changes requested):
라젤 피드백 반영 잘해주셨어요. 👍👍 
추가적으로 몇몇 코멘트 남겨두었는데, 확인 부탁드릴게요. 🙏
미션 진행하면서 잘 이해가 안되거나 어려운 부분있으면 언제든지 DM 주세요. 😉

RaZel713 committed:
refactor: synchronized 키워드 제거

RaZel713 committed:
refactor: 메서드 명 변경

- Repository 에서 id 값을 사용하여 찾거나 삭제하는 메서드 명을 ~ By(long id)라고 네이밍한다.

RaZel713 committed:
refactor: DTO 패키지 생성 및 분리

RaZel713 committed:
refactor: RestController를 사용해 컨트롤러 수정

- @Controller를 @RestController로 변경하여 JSON 응답을 명확히 한다.
- 메서드는 상태 코드에 따라 ResponseEntity 사용을 유지하거나 ResponseEntity 없이 객체 직접 반환한다.

RaZel713 committed:
refactor: Service 메서드명 간결화

- createReservation() 과 같이 호출 시 중복되는 단어인 Reservation()을 제거한다.

RaZel713 committed:
refactor: rowMapper를 메서드로 추출

RaZel713 committed:
refactor: 패키지 명을 대문자에서 소문자로 변경

RaZel713 committed:
refactor: 예약 삭제 및 예약 시간 삭제 기능에서 삭제 여부를 반환하도록 수정

- Repository가 삭제된 행 수 대신 삭제 성공 여부를 반환한다.
- 이처럼 Service 계층이 Repository의 세부 구현에 의존하지 않도록 개선한다.

RaZel713 committed:
refactor: 중복 예약 시간 검사 시 COUNT 대신 EXISTS 사용하여 쿼리 간결성 및 성능 개선

RaZel713 committed:
refactor: Repository에서 insert 후 객체 반환 대신 id 반환, 객체를 불러오는 로직을 Service로 이동

RaZel713 committed:
refactor: 에러 사인 상수 처리

RaZel713 committed:
refactor: 요청과 응답으로 DTO 반환

RaZel713 committed:
fix: 역직렬화 오류 해결을 위해 도메인 생성자에 @JsonCreator 추가

RaZel713 reviewed (commented):
안녕하세요~ 로키! 👋
빠르게 반영해서 리뷰 요청 드리려고 했는데,
DTO를 사용해서 리팩터링 하는 과정에서 생각하지 못한 부분에서 오류가 발생해서 (역직렬화 생성자 문제 - `@JsonCreator`)
고치느라 시간이 많이 걸렸네요 ㅠㅜ 🥲
너무 늦게 답글 보내 드려서 죄송합니다!

RaZel713 committed:
refactor: 미사용 파일 제거

RaZel713 committed:
refactor: 테스트 클래스 네이밍 변경

Rok93 commented on src/test/java/roomescape/reservation/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,177 @@
+
+    @DisplayName("육단계: 데이터 추가 / 삭제하기")
+    @Test
+    void step_six() {
```
5 ~ 8 단계 테스트 케이스는 Controller에 대한 테스트인 것 같은데, 별도의 테스트 클래스에 위치시키면 어떨까요?

RaZel713 replied:
해당 테스트 케이스들을 `Controller` 테스트 클래스로 이동하였습니다.

Rok93 commented on src/test/java/roomescape/reservation/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,177 @@
+        reservationRepository.insert(reservation);
+
+        // then
+        assertThat(reservationRepository.findAllReservations()).hasSize(1);
```
ReservationRepositoryTest 클래스는 `reservationRepository` 객체를 테스트하기위한 테스트 클래스인데요. 

테스트를 위한 fixture 혹은 검증을위한 then 절에 reservationRepository의 기능을 사용하는건 조금 찝찝하지 않을까요? 🤔 
given 혹은 then 절에서 사용하고있는 기능은 순수 reservationRepository의 기능이 아닌 별도의 신뢰할 수 있는 방식으로 제공되어야할 것 같다고 생각되는데  라젤은 어떻게 생각하시나요? 😃

RaZel713 replied:
RepositoryTest 클래스에 있던 테스트들이 
클래스의 네이밍과 부합하지 않는다고 판단하여, 
테스트 목적에 맞게 `ControllerTest`와 `ServiceTest`로 분리 하였습니다

Rok93 commented on src/test/java/roomescape/time/ReservationTimeApiTest.java:
```diff
@@ -0,0 +1,42 @@
+                .then().log().all()
+                .statusCode(200);
+    }
+}
```
File Changed 에서 가장 마지막에 종종  ⛔️와 같은 표시가 뜨는데 이 표시는 무슨 의미인지 고민해보신적있으신가요? 😃

[GitHub 파일 끝에 개행이 필요한 이유 (no newline at a end of file)](https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/)

RaZel713 replied:
페어에게 파일 끝에 개행을 넣는 것이 좋다는 이야기를 들었는데, 
이에 대해 자세히 알아본 적이 없었습니다!

리뷰어님께서 알려주셔서 새로운 사실을 알 수 있게 되었습니다!
감사합니다~😊

RaZel713 committed:
refactor: controller에서 serviceDTO를 생성해 넘기도록 수정하여 controller-service 간 양방향 의존성 제거

RaZel713 committed:
refactor: dto 패키지를 controller 패키지와 service 패키지로 각각 이동

RaZel713 committed:
refactor: 예약과 예약 시간의 생성자를 정적 팩토리 메서드로 변경

Rok93 commented on src/main/java/roomescape/domain/reservation/Reservation.java:
```diff
@@ -0,0 +1,56 @@
+    private final ReservationTime time;
+
+    @JsonCreator
+    public Reservation(Long id, String name, LocalDate date, ReservationTime time) {
```
> DTO를 사용해서 리팩터링 하는 과정에서 생각하지 못한 부분에서 오류가 발생해서 (역직렬화 생성자 문제 - @JsonCreator)
고치느라 시간이 많이 걸렸네요 ㅠㅜ 🥲
너무 늦게 답글 보내 드려서 죄송합니다!

어떤 이슈가 있으셨고 어떻게 해결하셨는지도 공유해주실 수 있을까요? 😃

Rok93 replied:
혹시 `@JsonCreator`를 붙이신 이유를 추측해보자면, 미션에서 주어진 테스트 코드가 통과하지 않아서? 라고 추측되는데요. 👀
맞으시다면 깨지는 테스트가 어떤 것을 테스트하고있는지, 또 요청이 들어갔을 때 어떤 응답이 나와야하는지 확인해보시면 좋을 것 같아요. 

원인을 찾으셔서 수정하신다면, 도메인 객체 내부의 `@JsonCreator` 혹은 `@JsonProperty` 애너테이션들은 제거해도 정상적으로 테스트가 수행될 것 같습니다. 😃

RaZel713 replied:
저장된 데이터 조회하는 테스트가 통과되지 않는 문제가 발생하였습니다!

직렬화된 데이터를 다시 자바 객체로 돌리는 과정(역직렬화)에서 해당 객체의 생성자가 필요하는데,
기본 생성자가 아닌 생성자를 2개 작성하게 되면서, 역직렬화 과정에서 오류가 발생하게 되었습니다. 

따라서 `@JsonCreator` 애너테이션을 활용하여 역직렬화 과정에서 사용할 생성자를 지정해 주면서 
해결했었습니다!

하지만 지금 코드를 확인해보니, 생성자를 하나로 두고 
정적 팩토리 메서드를 활용하여 데이터 베이스에 저장되지 않은 객체를 표현하고 있으므로, 
해당 애너테이션을 제거해도 문제가 발생하지 않으므로, 제거하였습니다!
 
<details>
<summary> 추가 학습) 생성자와 역직렬화의 관계 학습 </summary>

#### 생성자가 없는 경우
- 자동으로 기본 생성자를 만들어 역직렬화한다.
- Jackson은 이 기본 생성자 + setter 혹은 필드 접근 방식으로 객체를 생성한다.

#### 생성자가 하나인 경우
- Jackson이 해당 생성자를 사용한다.
  - 단, 파라미터 이름이 JSON key와 일치하거나, `@JsonProperty`로 지정되어 있어야 한다.

#### 사용자 정의 생성자가 둘 이상인 경우
- 어떤 생성자를 사용할지 Jackson이 혼란스럽다.
- 2가지 방법 중 하나를 택해서 사용해야 한다.
  - 하나에만 @JsonCreator 붙이고, 파라미터에 `@JsonProperty("필드명")` 붙여서 사용할 생성자를 명시적으로 지정해야 한다.
  - 기본 생성자를 하나 추가해서 Jackson이 그것을 사용할 수 있도록 해야한다.
  
</details>

Rok93 replied:
말씀하신 내용들도 맞지만  실제 API의 응답의 반환 타입은 `List<ReservationResponse>` 인데, 테스트 코드는 `Reservation` 타입으로 테스트하고있더라구요. 😃


타입이 달라서 문제가 될 수도 있겠더라구요. (물론 상태 네이밍이 모두 같아서 이슈가 없었을 수도 있었겠지만 ㅎㅎ.... DTO 객체는 record 클래스라서 언그하신 문제들이 없으셨을 것 같아요. 😉)

Rok93 reviewed (changes requested):
라젤 피드백 반영 잘해주셨어요! 👍💯
조금만 더 짚고 넘어가면 좋을부분이보여서, 코멘트 남겨두었어요. 😉
확인해서 반영 부탁드릴게요. 😃
잘 이해가 안되거나 어렵게느껴지는 코멘트는 언제든지 DM 주세요 .

RaZel713 committed:
refactor: 미사용 애너테이션 제거

RaZel713 committed:
test: controller 테스트 클래스 및 패키지 생성

RaZel713 committed:
test: service 테스트 클래스 및 패키지 생성

RaZel713 committed:
test: domain 테스트 클래스 및 패키지 생성

RaZel713 committed:
refactor: 메서드 분리

RaZel713 reviewed (commented):
로키! 좋은 피드백 많이 주셔서 학습하는데 큰 도움이 되었습니다! 
다음 레벨에도 리뷰어와 리뷰이 관계로 꼭 다시 만나고 싶습니다 😊

그동안 정말 감사했습니다!!🖐️

Rok93 reviewed (approved):
라젤 마지막까지 피드백 반영 잘해주셨어요. 👍👍 
제 피드백이 도움이 많이 되셨으면 좋겠네요. 🙏
방탈출 예약 관리 미션은 이만 종료하도록하게습니다. 🎉👏

고생많으셨고 남은 미션들도 화이팅입니다. 💪💪💪
